File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/persistence/JpaWishlistRepository.java
package com.zappymart.adapters.persistence;

import com.zappymart.application.ports.WishlistRepository;
import com.zappymart.domain.accounts.Wishlist;

import java.util.List;

public final class JpaWishlistRepository implements WishlistRepository {

    private final WishlistEntryRowRepository wishlistEntryRows;

    JpaWishlistRepository(WishlistEntryRowRepository wishlistEntryRows) {
        this.wishlistEntryRows = wishlistEntryRows;
    }

    @Override
    public Wishlist ofOwner(String ownerId) {
        List<String> productIds = wishlistEntryRows.findByOwnerIdOrderByPositionAsc(ownerId).stream()
                .map(row -> row.productId)
                .toList();
        return new Wishlist(ownerId, productIds);
    }

    @Override
    public Wishlist save(Wishlist wishlist) {
        wishlistEntryRows.deleteByOwnerId(wishlist.ownerId());
        wishlistEntryRows.flush();
        List<String> productIds = wishlist.productIds();
        for (int position = 0; position < productIds.size(); position++) {
            wishlistEntryRows.save(new WishlistEntryRow(wishlist.ownerId() + ":" + productIds.get(position),
                    wishlist.ownerId(), productIds.get(position), position));
        }
        return wishlist;
    }

    @Override
    public void delete(String ownerId) {
        wishlistEntryRows.deleteByOwnerId(ownerId);
    }
}

File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/persistence/JpaWishlistRepository.kt
package nl.zappymart.adapters.persistence

import nl.zappymart.adapters.persistence.entities.WishlistEntryEntity
import nl.zappymart.application.ports.WishlistRepository
import nl.zappymart.domain.accounts.Wishlist
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
@Transactional(readOnly = true)
class JpaWishlistRepository(private val entries: WishlistEntries) : WishlistRepository {

    override fun findByOwnerId(ownerId: String): Wishlist = Wishlist(
        ownerId,
        entries.findByOwnerIdOrderByPositionAsc(ownerId).map { entry -> entry.productId },
    )

    @Transactional
    override fun save(wishlist: Wishlist) {
        entries.deleteByOwnerId(wishlist.ownerId)
        entries.flush()
        wishlist.productIds.forEachIndexed { position, productId ->
            entries.save(WishlistEntryEntity(ownerId = wishlist.ownerId, productId = productId, position = position))
        }
    }

    @Transactional
    override fun delete(ownerId: String) {
        entries.deleteByOwnerId(ownerId)
    }
}
